scripts/cleanup_old_vms.py

```diff
@@ -217,75 +222,87 @@ def cleanup_vms(texts, max_hours=24, providers=None, prompt=True):
+    Args:
+        texts (list): List of regex strings to match with
+        max_hours (int): age limit for deletion
+        providers (list): List of provider keys to scan and cleanup
+        prompt (bool): Whether or not to prompt the user before deleting vms
+    Returns:
+        int: return code, 0 on success, otherwise raises exception
     """
     providers = providers or list_provider_keys()
     delta = datetime.timedelta(hours=int(max_hours))
-    vms_to_delete = defaultdict(set)
-    thread_queue = []
-    # precompile regexes
-    matchers = [re.compile(text, re.IGNORECASE) for text in texts]
+    # Compile regex, strip leading/trailing single quotes from cli arg
+    matchers = [re.compile(text.strip("'"), re.IGNORECASE) for text in texts]
     print('Matching VM names against the following case-insensitive strings: {}'.format(texts))
 
+    providers_to_scan = []
     for provider_key in providers:
         # check for cleanup boolean
         if not cfme_data['management_systems'][provider_key].get('cleanup', False):
             print('Skipping {}, cleanup map set to false or missing in yaml'.format(provider_key))
             continue
+        # check the provider is reachable
         ipaddress = cfme_data['management_systems'][provider_key].get('ipaddress')
         if ipaddress and not net.is_pingable(ipaddress):
             continue
-        thread = Thread(target=process_provider_vms,
-                        args=(provider_key, matchers, delta, vms_to_delete))
-        # Mark as daemon thread for easy-mode KeyboardInterrupt handling
-        thread.daemon = True
-        thread_queue.append(thread)
-        thread.start()
-
-    # Join the queued calls
-    for thread in thread_queue:
-        thread.join()
+        # passed the checks
+        providers_to_scan.append(provider_key)
+
+    scan_queue = Queue()
+    scan_failure_queue = Queue()
+    scan_proc_list = []
+    for provider_key in providers_to_scan:
+        scan_proc_list.append(
+            Process(target=scan_provider,
+                    args=(provider_key, matchers, delta, scan_queue, scan_failure_queue)))
+
+    for proc in scan_proc_list:
+        proc.start()
+    for proc in scan_proc_list:
+        proc.join()
+
+    # each item on scan_queue is a VmData tuple
+    vms_to_delete = []
+    while not scan_queue.empty():
+        vms_to_delete.extend(scan_queue.get())
 
     if vms_to_delete and prompt:
-        yesno = raw_input('Delete these VMs? [y/N]: ')
+        yesno = raw_input('Delete the-se VMs? [y/N]: ')
         if str(yesno).lower() != 'y':
             print('Exiting.')
             return 0
 
     if not vms_to_delete:
         print('No VMs to delete.')
 
-    thread_queue = []
-    for provider_key, vm_set in vms_to_delete.items():
-        provider_mgmt = get_mgmt(provider_key)
+    delete_queue = Queue()
+    delete_proc_list = []
+    for provider_key, vm_name, age in vms_to_delete:
```

Comment:
I just highlighted hte piece of code where you probably fork a lot of processes again